File: 22.cpp
class Solution {
public:
    void f(int opened, vector<string> &arr, string s, int n){
        int closed = s.length() - opened;
        if(n == opened){
            if(opened == closed){
                arr.push_back(s);
                return;
            }else{
                if(closed < opened){
                    f(opened, arr, s+')', n);
                    return;
                }
            }
        }
        else{
            if(opened > closed){
                f(opened, arr, s+')', n);
                f(opened+1, arr, s+'(', n);
                return;
            }else{
                f(opened+1, arr, s+'(', n);
            }
        }
        return;   
    }

    vector<string> generateParenthesis(int n) {
        vector<string> arr;
        f(0, arr, "",n);
        return arr;
    }
};

File: Python/22.py
class Solution:
    def f(self, sol, s,e,cs, n):
        if s == n and e == n:
            sol.append(cs)
        if s < n:
            self.f(sol, s+1, e, cs+'(', n)
        if e < n and e < s:
            self.f(sol, s, e+1, cs+')', n)

    def generateParenthesis(self, n: int) -> List[str]:
        sol = []
        self.f(sol, 0,0,'', n)
        return sol
